### ISSUE TITLE
App build succeeds with Pyinstaller version 4.3 but not 4.4 due to AttributeError: Module 'PyQt5' has no attribute '__version__'

### ISSUE THREAD

**Issue by fishbacp** [pyinstaller/pyinstaller#6293]
I'm using Python 3.9.5 and wanted to upgrade my Pyinstaller to the most recent version since I was encountering problems getting my app notarized. Everything worked fine with version 4.3, but building with 4.4 failed and resulted in a lengthy message ending with

AttributeError: Module 'PyQt5' has no attribute '__version__'

**Comment by bwoodsend** [comment]
Can you please fill out the [issue template](https://github.com/pyinstaller/pyinstaller/blob/develop/.github/ISSUE_TEMPLATE/bug_report.md). We can't help you if we don't know how it happened.

**Comment by rokm** [comment]
> AttributeError: Module 'PyQt5' has no attribute 'version'

Sounds like you have a partial installation of `PyQt5` in your environment, i.e., you probably removed `PyQt5` package but not `PyQt5-Qt5` and/or `PyQt5-sip`. Due to a regression, latest PyInstaller versions raise an error in that situation, but this will be fixed in upcoming release by #6141 (for `PyQt6` ) and #6114 (for `PyQt5`; although primary aim here was support for old conda version).

**Comment by fishbacp** [comment]
I had installed it and PyQt5 a few months ago when I was considering using Qt Designer for my GUI. (Now I just use tkinter.) Based upon your comment, I thought I'd try removing PyQt5 just to see what would happen. Now I was able to install Pyinstaller 4.4. 

I ran Pyinstaller with my spec file and it seemed to work out okay. However, the app wouldn't open upon double clicking. When I tried to run it from the command line, I obtained the following lengthy error message:

```
fishbacp@fishbacpK0ML85 Python_May_2021 % ./dist/My_Application_10_7.app/Contents/MacOS/My_Application_10_7
2021-10-11 13:34:52.300 My_Application_10_7[95597:4939036] *** Assertion failure in -[NSMenuItem initWithTitle:action:keyEquivalent:], NSMenuItem.m:454
2021-10-11 13:34:52.302 My_Application_10_7[95597:4939036] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Invalid parameter not satisfying: aString != nil'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff208c898b __exceptionPreprocess + 242
	1   libobjc.A.dylib                     0x00007fff20600d92 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff208f1af2 +[NSException raise:format:arguments:] + 88
	3   Foundation                          0x00007fff216a9a12 -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 191
	4   AppKit                              0x00007fff2303bad2 -[NSMenuItem initWithTitle:action:keyEquivalent:] + 281
	5   libtk8.6.dylib                      0x000000011cb31874 +[NSMenuItem(TKUtils) itemWithSubmenu:] + 67
	6   libtk8.6.dylib                      0x000000011cb2de93 -[TKApplication(TKMenu) tkSetMainMenu:] + 257
	7   libtk8.6.dylib                      0x000000011cb2f1ac TkpSetMainMenubar + 166
	8   libtk8.6.dylib                      0x000000011caae33b TkSetWindowMenuBar + 777
	9   libtk8.6.dylib                      0x000000011caa762e ConfigureFrame + 258
	10  libtk8.6.dylib                      0x000000011caa7250 FrameWidgetObjCmd + 545
	11  libtcl8.6.dylib                     0x000000011c9199f4 TclNRRunCallbacks + 79
	12  _tkinter.cpython-39-darwin.so       0x000000011c8eb9c0 Tkapp_Call + 480
	13  Python                              0x000000010d750ea3 method_vectorcall_VARARGS + 116
	14  Python                              0x000000010d83cc53 call_function + 164
	15  Python                              0x000000010d83a489 _PyEval_EvalFrameDefault + 45167
	16  Python                              0x000000010d745240 _PyFunction_Vectorcall + 192
	17  Python                              0x000000010d747e7a method_vectorcall + 160
	18  Python                              0x000000010d83cc53 call_function + 164
	19  Python                              0x000000010d83a526 _PyEval_EvalFrameDefault + 45324
	20  Python                              0x000000010d82e2d6 _PyEval_EvalCode + 403
	21  Python                              0x000000010d7452f8 _PyFunction_Vectorcall + 376
	22  Python                              0x000000010d747e7a method_vectorcall + 160
	23  Python                              0x000000010d83cc53 call_function + 164
	24  Python                              0x000000010d83a612 _PyEval_EvalFrameDefault + 45560
	25  Python                              0x000000010d82e2d6 _PyEval_EvalCode + 403
	26  Python                              0x000000010d82e13d PyEval_EvalCode + 72

```

**Comment by rokm** [comment]
Can you try with PyInstaller 4.5.1 instead of 4.4?

In 4.4, we tried to enable Apple event processing for `onedir` app bundles by default, but that didn't pan out and we reverted it in 4.5...

**Comment by fishbacp** [comment]
That did it! Thanks.

On a slightly different note, I see in 4.4 there is an option --codesign-identity to help with codesigning. Is this something I can include in my .spec file, and, if so, how. (I have a Developer ID Application.)

**Comment by rokm** [comment]
> On a slightly different note, I see in 4.4 there is an option --codesign-identity to help with codesigning. Is this something I can include in my .spec file, and, if so, how. (I have a Developer ID Application.)

Yes, the `--codesign-identity` switch controls the `codesign_identity` argument of `EXE()` in the spec file. You can add the `codesign_identity='MY_CODESIGN_ID'` argument to the `EXE` in an existing spec (in newly-generated ones, `codesign_identity=None` will be added by default unless you specify `--codesign-identity`). The specified `MY_CODESIGN_ID` string is then passed to the `codesign` utility via `-s` argument.

**Comment by fishbacp** [comment]
That's a really great feature! So I added that option to EXE() and the app bundle was created successfully. Then I entered 

`codesign -s "MyID" -v --deep --timestamp  -o runtime /Users/fishbacp/Desktop/dist/My_Application.app`

and was told the app was already signed, which I expected since I edited my .spec file. However, notarization failed for following reason:

"My_Application.zip/My_Application.app/Contents/Resources/sklearn/.dylibs/libomp.dylib"

is not signed and does not include a secure timestamp. 

When I was using trying to codesign without modifying my .spec file, I also ran into this problem with sklearn. Would this be an issue I should ask others about via the pyinstaller mailing list? I don't want to get off-topic here. 

Thanks.

**Comment by bwoodsend** [comment]
> Would this be an issue I should ask others about via the pyinstaller mailing list? I don't want to get off-topic here.

I'm fairly certain that no one on the mailing list knows how to get codesign and notarisation to work. There have been several posted questions on it but nobody was able to answer them.

**Comment by bwoodsend** [comment]
I believe that [this](https://gist.github.com/txoof/0636835d3cc65245c6288b2374799c43) is the best resource we've got on the topic. (At least the last time I pointed someone to it, they reported that it worked.)

**Comment by rokm** [comment]
> `codesign -s "MyID" -v --deep --timestamp -o runtime /Users/fishbacp/Desktop/dist/My_Application.app`

Does it help if you run this post-hoc codesign command with `--force`? I think it should re-sign everything again. If that solves the problem, then we somehow missed that file (e.g., it got smuggled in via `datas` as opposed to `binaries`...).

If it does not, then it could be that the dot character in `.dylibs` is throwing off the `codesign`. I think I recall someone mentioning something to that effect once...

**Comment by rokm** [comment]
Actually, that `sklearn/.dylibs` folder shouldn't be collected at all, because we currently go out of our way to collect shared libraries into top-level directory and rewrite references so that they point there. So there should also be a copy of `libomp.dylib` in top-level directory, and that `sklearn/.dylibs` folder can be removed altogether.

And our [`sklearn` hook](https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/9c3647cfc0c82d0781aed33748b95a9e41905eb7/src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-sklearn.py#L16) (or `collect_data_files` in general) needs to be patched not to collect `.dylib` files as data.

**Comment by rokm** [comment]
@fishbacp Patching your local copy of `[path-to-venv]/lib/python3.X/site-packages/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-sklearn.py` and changing the line:
```python
datas = collect_data_files('sklearn')
``` 
into
```python
datas = collect_data_files('sklearn', excludes=['**/*.dylib'])
```
should prevent collection of the problematic copy in `sklearn/.dylibs` directory. I think the app should still run, because it's the `libomp.dylib` copy collected in the top-level directory that actually matters.

**Comment by fishbacp** [comment]
I did run the codesign using --force but the notarization failed again for the same reasons as before. With regard to the dot character, I had found something related at https://github.com/pyinstaller/pyinstaller/wiki/Recipe-OSX-Code-Signing-Qt for when the app uses Qt. 

On my machine one of the .dylibs folders is labeled an "alias" and is in Contents\MacOS\sklearn but hidden (Shows with CMD-SHIFT-period)
The other one is Contents\Resources\sklearn and is also hidden.
